.github/actions/build-action/action.yml

```diff
@@ -0,0 +1,162 @@
+    description: 'Boolean: Run selected manual tests only'
+    required: false
+  ASAN_OPTIONS:
+    description: 'Options to use in asan builds'
+    required: false
+  UBSAN_OPTIONS:
+    description: 'Options to use in ubsan builds'
+    required: false
+  TSAN_OPTIONS:
+    description: 'Options to use in tsan builds'
+    required: false
+  NINJA_BUILD:
+    description: 'Boolean: Build with Ninja'
+    required: false
+  CMAKE_EXE:
+    description: 'Use a specific instance of CMake'
+    required: false
+outputs:
+  random-number:
+    description: "Random number"
+    value: ${{ steps.random-number-generator.outputs.random-id }}
+runs:
+  using: "composite"
+  steps:
+    - name: Dump-environment
+      shell: bash
+      run: |
+        env
+        set | grep =
+
+    - name: Install ccache
+      shell: bash
+      run: |
+        if ! type -a ccache
+        then
+          apt install ccache
+        fi
+
+    - name: Reset ccache stats
+      shell: bash
+      run: |
+          ccache -s
+          # Reset the stats so only this build counts
+          ccache -z
+
+    - name: Disk space
+      shell: bash
+      run: |
+        df -h
+        if [[ -d "${CACHE_DIR}" ]]
+        then
+            du -sh ${CACHE_DIR}
+        fi
+        if [[ -d "${CCACHE_DIR}" ]]
+        then
+            du -sh ${CCACHE_DIR}
+        fi
+        if [[ -d "${NIH_CACHE_ROOT}" ]]
+        then
+            find ${NIH_CACHE_ROOT} -maxdepth 3 \( -iname src -prune -o \
+              -type d -exec du -sh {} \; \)
+            # Post-condition: The cache will not be reused to build with a
+            # different compiler
+            if [[ "${CC}" =~ "gcc" ]]
+            then
+              echo Erasing NIH cache for clang
+              rm -rf "${NIH_CACHE_ROOT}"/*/Clang_*/ || true
+            elif [[ "${CC}" =~ "clang" ]]
+            then
+              echo Erasing NIH cache for gcc
+              rm -rf "${NIH_CACHE_ROOT}"/*/GNU_*/ || true
+            fi
+            # Free up space if the cache is being reused with a different
+            # generator from prior builds (ninja vs. makefiles).
+            # Post condition: Cache will not be reused with a different
+            # generator.
+            if [[ "${NINJA_BUILD}" == "true" && -d "${NIH_CACHE_ROOT}"/unix_makefiles && \
+              $( du -sm "${NIH_CACHE_ROOT}"/unix_makefiles | cut -f1 ) -gt 1000  ]]
+            then
+              echo Erasing NIH cache for makefiles
+              rm -rf "${NIH_CACHE_ROOT}"/unix_makefiles || true
+            elif [[ "${NINJA_BUILD}" == "false" && -d "${NIH_CACHE_ROOT}"/ninja && \
+              $( du -sm "${NIH_CACHE_ROOT}"/ninja | cut -f1 ) -gt 1000  ]]
+            then
+              echo Erasing NIH cache for ninja
+              rm -rf "${NIH_CACHE_ROOT}"/ninja || true
+            fi
+        fi
+
+    - name: Build and test
+      shell: bash
+      run: |
+          export CXX=${CC/gcc/g++}
+          export CXX=${CXX/clang/clang++}
```

Comment:
> If $CXX enters these lines as "clang++", won't it exit as "clang++++"? Did you mean to write ${CC/... on line 114?

No, because `CXX` is initialized on line 113, regardless of the value of `$CC`.
* `CC=gcc`
  * 113: `CXX=g++` (substitution)
  * 114: `CXX=g++` (no substitution)
* `CC=clang`
  * 113: `CXX=clang` (no substitution)
  * 114: `CXX=clang++` (substitution)

This is another sort of artifact of specifying the `matrix`. It's combinatoric, so it's non-trivial to set two variables together. If I did something like 
```
CC: [ "gcc", "clang" ]
CXX: [ "g++", "clang++" ]
```
Then I'd get 4 jobs
* gcc/g++
* gcc/clang++
* clang/g++
* clang/clang++
And it would take extra work to exclude the redundant ones. This let's me do what I want in one place with one variable.

Reply:
> it's non-trivial to set two variables together

This isn't true, and knowing the technique may open new ways to think about how to structure these actions (I'm looking at the relationship between `image` and `CMAKE_ARGS` too). Here's an [example](https://github.com/thejohnfreeman/hello-github-actions/blob/master/.github/workflows/matrix.yml), and it's [run](https://github.com/thejohnfreeman/hello-github-actions/runs/2729020953?check_suite_focus=true). The key part of the matrix is this:

```yml
matrix:
  compiler:
    - cc: gcc
      cxx: g++
    - cc: clang
      cxx: clang++
```

The key insight is that `matrix` is a dictionary of keys to arrays of values. Those values can be compound values like more dictionaries. Coupling related variables in a dictionary is a way to get them all in the matrix expression without generating excess or invalid configurations.

Reply:
In fact you used the technique [here](https://github.com/ripple/rippled/pull/3851/files#diff-d00d2382431454fdb50c1d36d99ec5836472a8831427e7e3f3a04385633cc695R38-R51).

Reply:
I had assumed that `include` and `exclude` were special cases. So you've taught me another new thing. Thanks! I went through and restructured the jobs and the action as much as possible to take advantage of compound values, including getting rid of the logic based on `image`.